File: frontend/src/services/productService.js
const API_URL = 'http://localhost:8080/api/products';

// Helper para manejar respuestas
const handleResponse = async (response) => {
    const contentType = response.headers.get('content-type');

    if (contentType && contentType.includes('application/json')) {
        const data = await response.json();

        if (!response.ok) {
            throw new Error(data.error || data.message || 'Error en la solicitud');
        }

        return data;
    }

    if (!response.ok) {
        if (response.status === 404) {
            throw new Error('Producto no encontrado');
        }
        if (response.status === 500) {
            throw new Error('Error del servidor');
        }
        throw new Error(`Error ${response.status}`);
    }

    return null;
};

const handleError = (error) => {
    if (error.message.includes('Failed to fetch')) {
        throw new Error('No se pudo conectar con el servidor');
    }
    throw error;
};

export const productService = {
    // Obtener todos los productos
    getAllProducts: async () => {
        try {
            const response = await fetch(`${API_URL}`, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                },
            });
            return await handleResponse(response);
        } catch (error) {
            handleError(error);
        }
    },

    // Obtener producto por ID
    getProductById: async (id) => {
        try {
            const response = await fetch(`${API_URL}/${id}`, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                },
            });
            return await handleResponse(response);
        } catch (error) {
            handleError(error);
        }
    },

    // Buscar por categoría
    getProductsByCategory: async (category) => {
        try {
            const response = await fetch(`${API_URL}/category/${category}`, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                },
            });
            return await handleResponse(response);
        } catch (error) {
            handleError(error);
        }
    },

    // Buscar por estado
    getProductsByStatus: async (status) => {
        try {
            const response = await fetch(`${API_URL}/status/${status}`, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                },
            });
            return await handleResponse(response);
        } catch (error) {
            handleError(error);
        }
    },

    // Buscar por nombre
    searchProducts: async (name) => {
        try {
            const response = await fetch(`${API_URL}/search?name=${encodeURIComponent(name)}`, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                },
            });
            return await handleResponse(response);
        } catch (error) {
            handleError(error);
        }
    },

    // Obtener categorías
    getCategories: async () => {
        try {
            const response = await fetch(`${API_URL}/categories`, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                },
            });
            return await handleResponse(response);
        } catch (error) {
            handleError(error);
        }
    },

    // ====== ENDPOINTS PROTEGIDOS (Requieren token) ======

    // Crear producto (Solo ADMIN)
    createProduct: async (productData, token) => {
        try {
            const response = await fetch(`${API_URL}`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${token}`,
                },
                body: JSON.stringify(productData),
            });
            return await handleResponse(response);
        } catch (error) {
            handleError(error);
        }
    },

    // Actualizar producto (Solo ADMIN)
    updateProduct: async (id, productData, token) => {
        try {
            const response = await fetch(`${API_URL}/${id}`, {
                method: 'PUT',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${token}`,
                },
                body: JSON.stringify(productData),
            });
            return await handleResponse(response);
        } catch (error) {
            handleError(error);
        }
    },

    // Eliminar producto (Solo ADMIN)
    deleteProduct: async (id, token) => {
        try {
            const response = await fetch(`${API_URL}/${id}`, {
                method: 'DELETE',
                headers: {
                    'Authorization': `Bearer ${token}`,
                },
            });
            return await handleResponse(response);
        } catch (error) {
            handleError(error);
        }
    },

    // Agregar stock (Solo ADMIN)
    addStock: async (id, quantity, token) => {
        try {
            const response = await fetch(`${API_URL}/${id}/stock/add?quantity=${quantity}`, {
                method: 'PATCH',
                headers: {
                    'Authorization': `Bearer ${token}`,
                },
            });
            return await handleResponse(response);
        } catch (error) {
            handleError(error);
        }
    },

    // Reducir stock (Solo ADMIN)
    reduceStock: async (id, quantity, token) => {
        try {
            const response = await fetch(`${API_URL}/${id}/stock/reduce?quantity=${quantity}`, {
                method: 'PATCH',
                headers: {
                    'Authorization': `Bearer ${token}`,
                },
            });
            return await handleResponse(response);
        } catch (error) {
            handleError(error);
        }
    },

    // Marcar como agotado (Solo ADMIN)
    markAsOutOfStock: async (id, token) => {
        try {
            const response = await fetch(`${API_URL}/${id}/out-of-stock`, {
                method: 'PATCH',
                headers: {
                    'Authorization': `Bearer ${token}`,
                },
            });
            return await handleResponse(response);
        } catch (error) {
            handleError(error);
        }
    },

    // Obtener productos con stock bajo (Solo ADMIN)
    getLowStockProducts: async (threshold, token) => {
        try {
            const response = await fetch(`${API_URL}/low-stock?threshold=${threshold}`, {
                method: 'GET',
                headers: {
                    'Authorization': `Bearer ${token}`,
                },
            });
            return await handleResponse(response);
        } catch (error) {
            handleError(error);
        }
    },

    // Obtener estadísticas (Solo ADMIN)
    getStats: async (token) => {
        try {
            const response = await fetch(`${API_URL}/stats`, {
                method: 'GET',
                headers: {
                    'Authorization': `Bearer ${token}`,
                },
            });
            return await handleResponse(response);
        } catch (error) {
            handleError(error);
        }
    },
};

File: backend/src/main/java/com/caprisoft/caprisoft/service/ProductService.java
package com.caprisoft.caprisoft.service;

import com.caprisoft.caprisoft.dto.ProductCreateDTO;
import com.caprisoft.caprisoft.dto.ProductDTO;
import com.caprisoft.caprisoft.dto.ProductUpdateDTO;
import com.caprisoft.caprisoft.entity.Product;
import com.caprisoft.caprisoft.entity.ProductCategory;
import com.caprisoft.caprisoft.entity.ProductStatus;
import com.caprisoft.caprisoft.exception.InsufficientStockException;
import com.caprisoft.caprisoft.exception.ProductNotFoundException;
import com.caprisoft.caprisoft.mapper.ProductMapper;
import com.caprisoft.caprisoft.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class ProductService {

    private final ProductRepository productRepository;
    private final ProductMapper productMapper;

    // ========== CREAR PRODUCTO ==========
    @Transactional
    public ProductDTO createProduct(ProductCreateDTO createDTO) {
        Product product = productMapper.toEntity(createDTO);
        Product savedProduct = productRepository.save(product);
        return productMapper.toDTO(savedProduct);
    }

    // ========== OBTENER TODOS LOS PRODUCTOS ==========
    @Transactional(readOnly = true)
    public List<ProductDTO> getAllProducts() {
        return productRepository.findByIsActiveTrue()
                .stream()
                .map(productMapper::toDTO)
                .collect(Collectors.toList());
    }

    // ========== OBTENER PRODUCTO POR ID ==========
    @Transactional(readOnly = true)
    public ProductDTO getProductById(Long id) {
        Product product = productRepository.findByIdAndIsActiveTrue(id)
                .orElseThrow(() -> new ProductNotFoundException(id));
        return productMapper.toDTO(product);
    }

    // ========== BUSCAR PRODUCTOS POR CATEGORÍA ==========
    @Transactional(readOnly = true)
    public List<ProductDTO> getProductsByCategory(ProductCategory category) {
        return productRepository.findByCategoryAndIsActiveTrue(category)
                .stream()
                .map(productMapper::toDTO)
                .collect(Collectors.toList());
    }

    // ========== BUSCAR PRODUCTOS POR ESTADO ==========
    @Transactional(readOnly = true)
    public List<ProductDTO> getProductsByStatus(ProductStatus status) {
        return productRepository.findByStatusAndIsActiveTrue(status)
                .stream()
                .map(productMapper::toDTO)
                .collect(Collectors.toList());
    }

    // ========== BUSCAR PRODUCTOS POR NOMBRE ==========
    @Transactional(readOnly = true)
    public List<ProductDTO> searchProductsByName(String name) {
        return productRepository.findByNameContainingIgnoreCaseAndIsActiveTrue(name)
                .stream()
                .map(productMapper::toDTO)
                .collect(Collectors.toList());
    }

    // ========== ACTUALIZAR PRODUCTO ==========
    @Transactional
    public ProductDTO updateProduct(Long id, ProductUpdateDTO updateDTO) {
        Product product = productRepository.findByIdAndIsActiveTrue(id)
                .orElseThrow(() -> new ProductNotFoundException(id));
        
        productMapper.updateEntityFromDTO(product, updateDTO);
        Product updatedProduct = productRepository.save(product);
        return productMapper.toDTO(updatedProduct);
    }

    // ========== ELIMINAR PRODUCTO (Soft Delete) ==========
    @Transactional
    public void deleteProduct(Long id) {
        Product product = productRepository.findByIdAndIsActiveTrue(id)
                .orElseThrow(() -> new ProductNotFoundException(id));
        
        product.setActive(false);
        productRepository.save(product);
    }

    // ========== ACTUALIZAR STOCK ==========
    @Transactional
    public ProductDTO updateStock(Long id, Integer quantity) {
        Product product = productRepository.findByIdAndIsActiveTrue(id)
                .orElseThrow(() -> new ProductNotFoundException(id));
        
        product.addStock(quantity);
        Product updatedProduct = productRepository.save(product);
        return productMapper.toDTO(updatedProduct);
    }

    // ========== REDUCIR STOCK ==========
    @Transactional
    public ProductDTO reduceStock(Long id, Integer quantity) {
        Product product = productRepository.findByIdAndIsActiveTrue(id)
                .orElseThrow(() -> new ProductNotFoundException(id));
        
        if (!product.hasStock(quantity)) {
            throw new InsufficientStockException(id, quantity, product.getStock());
        }
        
        product.reduceStock(quantity);
        Product updatedProduct = productRepository.save(product);
        return productMapper.toDTO(updatedProduct);
    }

    // ========== MARCAR COMO AGOTADO ==========
    @Transactional
    public ProductDTO markAsOutOfStock(Long id) {
        Product product = productRepository.findByIdAndIsActiveTrue(id)
                .orElseThrow(() -> new ProductNotFoundException(id));
        
        product.setStatus(ProductStatus.AGOTADO);
        Product updatedProduct = productRepository.save(product);
        return productMapper.toDTO(updatedProduct);
    }

    // ========== OBTENER PRODUCTOS CON STOCK BAJO ==========
    @Transactional(readOnly = true)
    public List<ProductDTO> getLowStockProducts(Integer threshold) {
        return productRepository.findLowStockProducts(threshold)
                .stream()
                .map(productMapper::toDTO)
                .collect(Collectors.toList());
    }

    // ========== ESTADÍSTICAS ==========
    @Transactional(readOnly = true)
    public Long countAvailableProducts() {
        return productRepository.countByStatusAndIsActiveTrue(ProductStatus.DISPONIBLE);
    }

    @Transactional(readOnly = true)
    public Long countOutOfStockProducts() {
        return productRepository.countByStatusAndIsActiveTrue(ProductStatus.AGOTADO);
    }
}